src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersReservedSlotsBehavior.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Connection;
+using Stratis.Bitcoin.P2P;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.P2P.Protocol;
+using Stratis.Bitcoin.P2P.Protocol.Behaviors;
+using Stratis.Bitcoin.P2P.Protocol.Payloads;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.Behaviors
+{
+    public class ProvenHeadersReservedSlotsBehavior : NetworkPeerBehavior
+    {
+        private readonly IConnectionManager connectionManager;
+        private readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        public ProvenHeadersReservedSlotsBehavior(
+            IConnectionManager connectionManager,
+            ILoggerFactory loggerFactory)
+        {
+            this.connectionManager = connectionManager;
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        /// <summary>
+        /// Processes and incoming message from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="message">Received message to process.</param>
+        protected async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
```

Comment:
Ah one more thing please look at `BlockStoreBehavior` I think we need to put a try catch for `OperationCanceledException`, this might be thrown when the node just disconnects

Reply:
As discussed, should happen only if you send something to node, so not needed.